### PR TITLE
G-26 | Handle replay attacks 

### PR DESCRIPTION
**Pull request by Saisree9**

### CODE REVIEW

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/RequestValidator.java  1
- src/test/java/in/projecteka/consentmanager/common/RequestValidatorTest.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java  6
- src/test/java/in/projecteka/consentmanager/link/discovery/DiscoveryUserJourneyTest.java  5
- src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequestController.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5826269&bid=19074774)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/RequestValidator.java  1
- src/test/java/in/projecteka/consentmanager/common/RequestValidatorTest.java  1
         

Clones added
============
- src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequestController.java  2
- src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java  6
- src/test/java/in/projecteka/consentmanager/link/discovery/DiscoveryUserJourneyTest.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5826269&bid=19074774)

File: src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java [hunk 1]
```diff
@@ -71,8 +73,8 @@
 import static in.projecteka.consentmanager.link.link.TestBuilders.string;
 import static in.projecteka.consentmanager.link.link.TestBuilders.user;
 import static java.util.List.of;
-import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.ArgumentMatchers.*;
```
Comment (Nexengineer) [inline src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java:76]: Remove wildcard import

Comment (imaditiagg) [reply src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java]: Done!

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/common/RequestValidator.java  1
- src/test/java/in/projecteka/consentmanager/common/RequestValidatorTest.java  1
         

Clones added
============
- src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequestController.java  2
- src/test/java/in/projecteka/consentmanager/link/discovery/DiscoveryUserJourneyTest.java  5
- src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5826269&bid=19074774)

File: src/main/java/in/projecteka/consentmanager/common/RequestValidator.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import java.time.format.DateTimeFormatter;
+
+@AllArgsConstructor
+public class RequestValidator {
+    private final CacheAdapter<String, String> cacheForReplayAttack;
+
+    public Mono<Boolean> validate(String requestId, String timestamp) {
+        return isRequestIdPresent(requestId)
+                .flatMap(isRequestIdPresent ->
+                        Mono.just(!isRequestIdPresent && isRequestIdValidInGivenTimestamp(timestamp)));
```
Comment (Apoorva-GA) [inline src/main/java/in/projecteka/consentmanager/common/RequestValidator.java:19]: You can do
```
return isRequestIdPresent(maybeRequest)
                .flatMap(result -> error(tooManyRequests()))
                .then(isRequestIdValidInGivenTimestamp(maybeRequest));
```

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/test/java/in/projecteka/consentmanager/common/RequestValidatorTest.java  1
- src/main/java/in/projecteka/consentmanager/common/RequestValidator.java  2
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/discovery/DiscoveryUserJourneyTest.java  3
- src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequestController.java  2
- src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5826269&bid=19074774)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/test/java/in/projecteka/consentmanager/common/RequestValidatorTest.java  1
- src/main/java/in/projecteka/consentmanager/common/RequestValidator.java  2
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/discovery/DiscoveryUserJourneyTest.java  3
- src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java  4
- src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequestController.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5826269&bid=19074774)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/test/java/in/projecteka/consentmanager/common/RequestValidatorTest.java  1
- src/main/java/in/projecteka/consentmanager/common/RequestValidator.java  2
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/link/discovery/DiscoveryUserJourneyTest.java  3
- src/main/java/in/projecteka/consentmanager/dataflow/DataFlowRequestController.java  2
- src/test/java/in/projecteka/consentmanager/link/link/LinkUserJourneyTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5826269&bid=19074774)

**Pull request merged**